### ISSUE TITLE
Have full control over the User-Agent

### ISSUE THREAD

**Issue by daladim** [hyperium/tonic#970]
### Motivation

The User-Agent header is populated, by default, with `tonic/x.y.z` (x.y.z. being the current tonic version number) \
There is a [user_agent](https://docs.rs/tonic/latest/tonic/transport/channel/struct.Endpoint.html#method.user_agent) method to somehow customize this by adding a prefix to this `tonic/x.y.z` part.

### Proposal

In some use cases, users could want to _totally_ customize the user-agent. Not suffixing with `tonic/x.y.z` would probably be a good idea.

I'm not sure why this suffix has been made compulsory anyway (but maybe I'm missing something? Is this required by some spec? I feel that any server _requiring_ to know the client version is probably poorly designed)

Thanks

**Comment by LucioFranco** [comment]
The endpoint code will be going away in future releases, so I am adverse to making changes like this right now. That said, you can always wrap your Client in a tower service and use https://docs.rs/tower-http/latest/tower_http/trait.ServiceBuilderExt.html#tymethod.override_request_header to set the user agent header with what you 'd like.

**Comment by stan-irl** [comment]
@LucioFranco isnt tonic transport lower level than tower middleware? wont tonic just override the tower value?

**Comment by stan-irl** [comment]
I just confirmed that behaviour with a test. the tonic transport overrides whatever value is set in the middleware layer

**Comment by roblabla** [comment]
Would you accept a PR adding a `dont_set_tonic_version_in_user_agent` (bikeshed pending) method on `Channel` to fix this issue?

**Comment by clouedoc** [comment]
Interested in a work-around as well. 

Some ideas in the meantime:

- forking the crate
- patching the crate (is it possible to patch a crate in `build.rs` in Rust? Kind of like JavaScript does with `patch-package`: https://www.npmjs.com/package/patch-package

I'll report if I make some progress

**Comment by clouedoc** [comment]
I forked tonic and patched here: https://github.com/clouedoc/tonic

You'll find the patch in the `patch-user-agent` branch.

My fork targets version 0.12.3.

Here's how to use it in your Cargo.toml:

```cargo
[dependencies]
prost = "0.13.4"
tonic = { git = "https://github.com/clouedoc/tonic.git", branch = "patch-user-agent", features = ["default", "channel", "server", "gzip"] }
# note: uncomment this when doing local development of tonic
#tonic = { version = "0.12.3", features = ["default", "channel", "server", "gzip"] }
tokio = { version = "1.0", features = ["macros", "rt-multi-thread"] }

# note: uncommit this when doing local development of tonic
#[patch.crates-io]
#tonic = { path = "../tonic/tonic" }
```

## Example

```rs
    let channel = Channel::from_static(addr).user_agent("my_custom_user_agent").unwrap().connect().await?; // is there a way to remove the unwrap here?

    let mut client = MyClient::with_interceptor(channel, |mut req: tonic::Request<()>| {
        req.metadata_mut().insert("Authorization", "XXXXXXXXXXXX".try_into().unwrap());
        Ok(req)
    });
```

Do note that the interceptor is not necessary, I'm just using it for something else.

## Future direction

It looks like changes to Endpoint are not accepted because the code is undergoing a rewrite?

I'd say that the best would be to wait for the 0.13.0 release and complain again.

**Comment by roblabla** [comment]
I'm not sure why this issue was closed, the linked PR does not, in fact, allow for full control over the User-Agent string. Reading the code, it looks like it appends the tonic UA to the User-Agent string.

This issue is **specifically** about avoiding this. Can we get this reopened?

**Comment by LucioFranco** [comment]
Yeah that is totally fair, I must have misread the full problem there. I will reopen the issue. PRs are welcome to solve this.

**Comment by i18nsite** [comment]
fix by https://github.com/hyperium/tonic/pull/2334/files
